---
title: "Identifying differentially expressed genes of gdT cells under IL-21 stimulation"
author: "Dan Griffiths"
date: "2023-12-21"
output:
  html_document: default
---

### Load libraries

```{r, message = FALSE}
library(limma)
library(Glimma)
library(tidyverse)
library(DESeq2)
library(org.Hs.eg.db)
library(pheatmap)
library(EnhancedVolcano)
library(ComplexHeatmap)
library(gplots)
library(msigdbr)
library(clusterProfiler)
library(enrichplot)
library(DOSE)
```

### Import data

```{r}
counts <- read_rds(file = "counts_matrix_raw.rds")
coldata <- read_rds(file = "coldata_dataframe.rds")
identical(colnames(counts), coldata$sample)
```

### Discussion of the project

This RNASeq dataset is to complement other assays conducting in determining the role of IL-21 in stimulating gd T cells to induce anti-inflammatory IL-10.

"When cultured in the presence of IL-21, Vγ9/Vδ2 T-cells acquired the ability to induce expression of the immunoregulatory cytokine IL-10 in both naïve and memory CD4+ T-cells..." (M Eberl, Unpublished)

This dataset contains PBMC extractions from three healthy donors. The gd T cells have been isolated and cultured under various stimuli: Unstimulated, IL-21, IL-2, IL-15.

While valuable information can be gleaned from this dataset, it is essential to acknowledge its limitations, given its small cohort size (n = 3). Preventing further, extensive analysis is the observation that one of the healthy donors (D3) diverges from the pattern observed in the other two donors. The unique gene expression pattern in D3 poses a challenge in discerning whether this deviation is attributable to technical or biological variation, given the cohort's limited size. The issues identified in D3 are detailed in the supplementary section of this markdown. Despite the absence of a clear rationale for excluding D3, it has been retained in the analysis. While it may have potentially suppressed the differential expression of certain genes, we anticipate that this will have limited effect as we focus on genes that exhibit the most differential expression between groups. Despite potential confounding from D3, we expect to identify the most differentially expressed genes. Our particular interest lies in genes related to cell-cell signaling, protein expression, and immunity; therefore, our analysis and presentation of data will emphasise this aspect of the data.

Given the challenges associated with the dataset, we place limited emphasis on analyses beyond differential expression. Brief pathway analyses have been conducted and are presented in the supplementary section of this document.

This dataset, along with other assays included in this project, indicates a mechanistic change in gd T cells during IL-21 stimulation but not in IL-2 and IL-15 stimulation. Therefore, the objective is to conduct a Wald test analysis of the log fold change of genes, comparing IL-21 vs IL-2 and IL-21 vs IL-15, with a specific focus on up-regulated genes in IL-21. Our focus on IL-21 up-regulated genes stems from our expectation of an active role of gd T cells in mechanistic interactions with CD4 T and CD4 T naive cells.

We aim to present a table of differentially expressed genes and visualize these results through a volcano plot. In particular, we are interested in genes related to immunity and cell signaling, and these immune-related genes will be appropriately labeled. This approach facilitates a discussion of differentially expressed genes with a focus on immune-related aspects, specifically CXCL13 and TNFRS8 (CD30).

The data pre-processing was conducted using NF-Core RNASeq nextflow package, and a count table produced using RSubread package. Full configuration and version information will be declared before publication.

### Rename gene list

```{r, message=FALSE}
gene.list <- data.frame(entrez.id = rownames(counts))
gene.list <- gene.list %>%
  dplyr::mutate(
    SYMBOL = mapIds(org.Hs.eg.db, keys = gene.list$entrez.id, column = "SYMBOL", keytype = "ENTREZID"),
    GENETYPE = mapIds(org.Hs.eg.db, keys = gene.list$entrez.id, column = "GENETYPE", keytype = "ENTREZID"),
    ENSEMBL = mapIds(org.Hs.eg.db, keys = gene.list$entrez.id, column = "ENSEMBL", keytype = "ENTREZID")
  )

# Rename to SYMBOL
filtered.gene.list <- gene.list %>%
  dplyr::filter(
    !is.na(ENSEMBL),
    !duplicated(ENSEMBL))

# Rename gene IDs from ensembl to symbol
counts <- counts[rownames(counts) %in% filtered.gene.list$entrez.id, ]
rownames(counts) <- filtered.gene.list$SYMBOL[match(rownames(counts), filtered.gene.list$entrez.id)]
```

### Generate DESeq2 object

```{r, message=FALSE, warning=FALSE}
dds <- DESeqDataSetFromMatrix(counts,coldata,design = ~healthy_donor + condition)
dds <- estimateSizeFactors(dds)
idx <- rowSums(counts(dds, normalized=TRUE) >= 10 ) >= 3
dds <- dds[idx,]
dds <- DESeq(dds)
vst <- varianceStabilizingTransformation(dds)
vsn::meanSdPlot(assay(vst))
plotPCA(vst)
sizeFactors(dds)
normalized_counts <- counts(dds, normalized=TRUE)
```

### DESeq2 results - Generalised linear model (Wald test)

```{r}
res.il15 <- results(dds,contrast = c("condition","IL21","IL15"))
res.sig.il15 <- subset(res.il15, padj < 0.05)
lfc.il15 <- res.sig.il15[ order(res.sig.il15$log2FoldChange, decreasing=TRUE), ]
summary(lfc.il15)
lfc.il15
res.il2 <- results(dds,contrast = c("condition","IL21","IL2"))
res.sig.il2 <- subset(res.il2, padj < 0.05)
lfc.il2 <- res.sig.il2[ order(res.sig.il2$log2FoldChange, decreasing=TRUE), ]
summary(lfc.il2)
lfc.il2

lfc.il15.df <- lfc.il15 %>% as.data.frame() %>% rownames_to_column(.,var = 'GENE_ID')
lfc.il2.df <- lfc.il2 %>% as.data.frame() %>% rownames_to_column(.,var = 'GENE_ID')
write_csv(lfc.il15.df, "lfc_il15_results.csv")
write_csv(lfc.il2.df, "lfc_il2_results.csv")

```

### Venn diagram - differentially expressed genes expressed across both tests 

```{r,fig.height=5,fig.width=5}
de.il2 <- subset(res.il2, abs(log2FoldChange) > 2.5 & -log10(padj) > -log10(0.05)) %>% rownames()
de.il15 <- subset(res.il15, abs(log2FoldChange) > 2.5 & -log10(padj) > -log10(0.05)) %>% rownames()
venn(list("IL15vsIL21" = de.il2, "IL2vsIL21" = de.il15), )
common.de.genes <- Reduce(intersect, list(de.il2, de.il15))
common.de.genes
```


```{r}
# get descriptors for the genes that are present in the IL-2, IL-15 comparisons (common.de.genes)

common.gene.df <- data.frame(og.genes = common.de.genes)
common.gene.list <- common.gene.df %>%
  dplyr::mutate(
    ONTOLOGY = mapIds(org.Hs.eg.db, keys = common.gene.df$og.genes, column = "ONTOLOGY", keytype = "SYMBOL"),
    GENETYPE = mapIds(org.Hs.eg.db, keys = common.gene.df$og.genes, column = "GENETYPE", keytype = "SYMBOL"),
    EVIDENCE = mapIds(org.Hs.eg.db, keys = common.gene.df$og.genes, column = "EVIDENCEALL", keytype = "SYMBOL"),
    GENE_NAME = mapIds(org.Hs.eg.db, keys = common.gene.df$og.genes, column = "GENENAME", keytype = "SYMBOL"),
    ENSEMBL = mapIds(org.Hs.eg.db, keys = common.gene.df$og.genes, column = "ENSEMBL", keytype = "SYMBOL")
  )
# select significantly differentially expressed genes that are shared between both statistical pairwise comparisons, that are part of a biological process, and that are supported by experimental evidence
cgl.evidence <- common.gene.list %>% dplyr::filter(ONTOLOGY == "BP") %>% dplyr::filter(!EVIDENCE %in% c("IBA","IBD","IKR","IRD"))
cgl.full <- common.gene.list

my_table <- nice_table(cgl.evidence)

print(my_table, preview = "docx")
# pdf("~/Desktop/table.pdf")
# print(my_table)
# dev.off()

```

### Volcano plot

Finally, taking the list of informative genes that we have obtained from the previous chapter, we show a volcano plot of differentially expressed genes. To note, that the padj has been left more inclusive, at 0.07, rather than the typical 0.05. This is due some observations not showing as classically statistically significant, but still being of note, as there is a substantial log2FoldChange.

```{r,fig.height=10,fig.width=10}

# Set fold change and p-value cutoffs
fold_cutoff <- 2.5
pvalue_cutoff <- 0.05

label_fold_cutoff <- 6
label_pvalue_cutoff <- 0.1

# Filter data based on cutoffs
res.il2.labels <- subset(res.il2, abs(log2FoldChange) > label_fold_cutoff & -log10(padj) > -log10(label_pvalue_cutoff))
res.il15.labels <- subset(res.il15, abs(log2FoldChange) > label_fold_cutoff & -log10(padj) > -log10(label_pvalue_cutoff))

res.il2.labels <- unique(paste0(c(rownames(res.il2.labels), "CXCL13", "IGFBP4", "TNFRSF8", "OSM", "IFNG", "DUSP6", "LTA", "SOCS2", "CSF2", "FCRL2")))
res.il15.labels <- unique(paste0(c(rownames(res.il15.labels), "CXCL13", "IGFBP4", "TNFRSF8", "OSM", "IFNG", "DUSP6", "LTA", "SOCS2", "CSF2", "FCRL2")))

filtered_data_intersect <- intersect(res.il2.labels, res.il15.labels)

EnhancedVolcano(res.il2,
    lab = rownames(res.il2),
    selectLab = filtered_data_intersect,
    drawConnectors = T,
    x = 'log2FoldChange',
    y = 'padj',
    title = 'IL21 versus IL2',
    pCutoff = pvalue_cutoff,
    FCcutoff = fold_cutoff,
    pointSize = 3.0,
    boxedLabels = TRUE,
    labSize = 6.0)

EnhancedVolcano(res.il15,
    lab = rownames(res.il15),
    selectLab = filtered_data_intersect,
    drawConnectors = T,
    x = 'log2FoldChange',
    y = 'padj',
    title = 'IL21 versus IL15',
    pCutoff = pvalue_cutoff,
    FCcutoff = fold_cutoff,
    pointSize = 3.0,
    boxedLabels = TRUE,
    labSize = 6.0)

```

## Supplementary analysis

Many other types of exploratory analysis was conducted on this dataset, but as mentioned earlier in the document, a small cohort size limits the extent further statistical methods can be applied.

For reference, a variance heatmap and a pathway analysis has been conducted on this dataset and can be seen below. The variance heatmap shows the extent to which D3 does not follow the same pattern of gene expression as the other healthy donor samples. The pathway analysis shows some indication of immune-related signalling changes, but the current dataset appears to be underpowered to draw significant conclusions from the data.

### Total variance heatmap

```{r,fig.height=10,fig.width=10}
vst_data <- assay(vst)
topVarGenes <- order(rowVars(vst_data), decreasing = TRUE)
mat  <- vst_data[topVarGenes[1:20], ]
# mat  <- vst_data[topVarGenes, ]
mat  <- mat - rowMeans(mat)

anno <- as.data.frame(colData(dds))
annocols1 <- paste(c(unique(anno$condition)))
annocols2 <- paste(c(unique(anno$healthy_donor)))
vip.cols1 <- RColorBrewer::brewer.pal(name = "Dark2", n = length(annocols1))
vip.cols2 <- RColorBrewer::brewer.pal(name = "Set2", n = length(annocols2))
names(vip.cols1) = sort(unique(anno$condition))
names(vip.cols2) = sort(unique(anno$healthy_donor))

HA <- columnAnnotation(df = anno %>% dplyr::select(c("condition","healthy_donor")), col = list(condition=vip.cols1,healthy_donor=vip.cols2))

Heatmap(mat, 
        name = "Relative variance",
        column_title = "Filtered counts matrix \n Variance stabilised expression heatmap \n Gene expression relative to the mean", 
        top_annotation = HA, 
        cluster_columns = TRUE, 
        #column_split = anno$healthy_donor,
        cluster_rows = TRUE
        )

```

### Brief pathway analysis using GO database

```{r}
# reading in data from deseq2
df <- res.il2
df <- data.frame(df)

# we want the log2 fold change 
original_gene_list <- df %>% rownames_to_column() %>% dplyr::select(c("rowname","log2FoldChange"))

# name the vector
names(original_gene_list) <- df$X

# omit any NA values 
gene_list<-na.omit(original_gene_list)

geneList = df %>% dplyr::select("log2FoldChange")
gene_list_ordered <- gene_list[order(gene_list[,2], decreasing = TRUE), ]
rownames(gene_list_ordered) <- gene_list_ordered[,1]
gene_list_ordered <- gene_list_ordered[,2]
names(gene_list_ordered) <- rownames(df)
```

```{r}
gse <- gseGO(geneList=gene_list_ordered,
             ont ="ALL", 
             keyType = "SYMBOL", 
             minGSSize = 3, 
             maxGSSize = 800, 
             pvalueCutoff = 0.05, 
             verbose = TRUE, 
             OrgDb = org.Hs.eg.db,
             pAdjustMethod = "none")
```

```{r,fig.height=15}
dotplot(gse, showCategory=20
        , split=".sign") +
  facet_grid(.~.sign)
```

```{r}
retrieved <- AnnotationDbi::select(org.Hs.eg.db, keytype="GOALL", keys=c("GO:0042101"), columns="SYMBOL")
head(retrieved)
retrieved <- unique(retrieved$SYMBOL)
```

### Further filtering and selection of results tables 
```{r}
# show dataframe with filtered higher baseMeans 
# although is difficult to interpret as it can be relative to other genes, and does not take into account transcript length
# lfc.il15.df %>% dplyr::filter(baseMean > 50, abs(log2FoldChange) > 2, padj < 0.07)


# metadata(res.il15)$filterThreshold
# 
# as_tibble(metadata(res.il15)$filterNumRej) %>%
#   ggplot(aes(x = theta, y = numRej)) +
#   geom_point() +
#   geom_vline(xintercept = 0.484,
#              color = 'red')
# hist(res.il15$pvalue)
```

## R session info

```{r}
sessionInfo()
```
