# Other resources:

-  Config files
https://zenodo.org/records/12790718

-  Github pages
https://dgrfs.github.io/RNASEQ_IL10_2016/

-  Published paper
https://doi.org/10.1093/discim/kyae008

-  General RNASeq processing pipeline for raw fastq files
https://github.com/dgrfs/NFCore_RNASeq_Cardiff_University_Hawk

-  Raw data --> 

ArrayExpress Accession: E-MTAB-13875

https://www.ebi.ac.uk/biostudies/arrayexpress/studies/E-MTAB-13875

Data can be downloaded here:

https://www.ebi.ac.uk/ena/browser/view/PRJEB73583

These samples were sequenced twice - we essentially have sequencing technical replicates (named run 35 and run 36). For the analysis, both sequencing runs are shown, and differential expression is applied to the two sequencing runs. There was no benefit to combining the data, and there are always concerns in terms of batch effect by combining data without correction. Therefore, the first sequencing run (named sequencing run 35 by the sequencing hub, and referenced as such throughout the analysis). As run 35 was used for the primary analysis, this is the data that is publicly available. In the sequencing run, IL-7 condition was also analysed; however, in the paper (and public raw data), the IL-7 condition was not pertinent, and so was not used in the final analysis. The IL-7 condition is shown in the markdowns but was removed for the analysis that was used for the published paper.

