---
title: "Differential Expression - RNA Sequencing data"
author: "Daniel D Griffiths"
date: "`r Sys.Date()`"
output: 
  html_document:
    toc: true
    toc_float: false
    highlight: tango
    includes:
      in_header: header.html
    code_folding: show
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Discussion of the project
This RNASeq dataset is to complement other assays conducting in determining the role of IL-21 in stimulating gd T cells to induce anti-inflammatory IL-10.

"When cultured in the presence of IL-21, Vγ9/Vδ2 T-cells acquired the ability to induce expression of the immunoregulatory cytokine IL-10 in both naïve and memory CD4+ T-cells..." (M Eberl, Unpublished)

This dataset contains PBMC extractions from three healthy donors. The gd T cells have been isolated and cultured under various stimuli: Unstimulated, IL-21, IL-2, IL-15.

While valuable information can be gleaned from this dataset, it is essential to acknowledge its limitations, given its small cohort size (n = 3). Preventing further, extensive analysis is the observation that one of the healthy donors (D3) diverges from the pattern observed in the other two donors. The unique gene expression pattern in D3 poses a challenge in discerning whether this deviation is attributable to technical or biological variation, given the cohort's limited size. The issues identified in D3 are detailed in the supplementary section of this markdown. Despite the absence of a clear rationale for excluding D3, it has been retained in the analysis. While it may have potentially suppressed the differential expression of certain genes, we anticipate that this will have limited effect as we focus on genes that exhibit the most differential expression between groups. Despite potential confounding from D3, we expect to identify the most differentially expressed genes. Our particular interest lies in genes related to cell-cell signaling, protein expression, and immunity; therefore, our analysis and presentation of data will emphasise this aspect of the data.

Given the challenges associated with the dataset, we place limited emphasis on analyses beyond differential expression. Brief pathway analyses have been conducted and are presented in the supplementary section of this document.

This dataset, along with other assays included in this project, indicates a mechanistic change in gd T cells during IL-21 stimulation but not in IL-2 and IL-15 stimulation. Therefore, the objective is to conduct a Wald test analysis of the log fold change of genes, comparing IL-21 vs IL-2 and IL-21 vs IL-15, with a specific focus on up-regulated genes in IL-21. Our focus on IL-21 up-regulated genes stems from our expectation of an active role of gd T cells in mechanistic interactions with CD4 T and CD4 T naive cells.

We aim to present a table of differentially expressed genes and visualize these results through a volcano plot. In particular, we are interested in genes related to immunity and cell signaling, and these immune-related genes will be appropriately labeled. This approach facilitates a discussion of differentially expressed genes with a focus on immune-related aspects, specifically CXCL13 and TNFRS8 (CD30).

The data pre-processing was conducted using NF-Core RNASeq nextflow package, and a count table produced using RSubread package. Full configuration and version information will be declared before publication.



# Load libraries

```{r message=FALSE, warning=FALSE}
library(limma)
library(Glimma)
library(tidyverse)
library(DESeq2)
library(org.Hs.eg.db)
library(pheatmap)
library(EnhancedVolcano)
library(ComplexHeatmap)
library(gplots)
library(msigdbr)
library(clusterProfiler)
library(enrichplot)
library(DOSE)
library(gridExtra)
library(grid)
```

# Motivation

-   There were two sequencing runs conducted on the same samples (same library pool re-sequenced)

-   Get log fold change volcanoes for both sequencing runs

-   Decide if these sequencing runs should be combined

# Set up enivronment

## Preferences

```{r}
set.seed(42)
```

## Load counts

```{r}
# These counts were generated on the HPC using RSubRead FeatureCounts package
load("source_data/FeatCounts_Rsub_run35.RData")
counts_run35 <- counts$counts
load("source_data/FeatCounts_Rsub_run36.RData")
counts_run36 <- counts$counts
head(counts_run35[1:4,1:4])
```

# Run 35

```{r}
counts <- counts_run35
```

## Coldata

```{r}
coldata <- tibble::tribble(
  ~SeqID, ~healthy_donor, ~condition,
  "MERNA1", "D1", "IL2",
  "MERNA2", "D1", "IL15",
  "MERNA3", "D1", "IL21",
  "MERNA4", "D1", "IL7",
  "MERNA5", "D1", "UST",
  "MERNA6", "D2", "IL2",
  "MERNA7", "D2", "IL15",
  "MERNA8", "D2", "IL21",
  "MERNA9", "D2", "IL7",
  "MERNA10", "D2", "UST",
  "MERNA11", "D3", "IL2",
  "MERNA12", "D3", "IL15",
  "MERNA13", "D3", "IL21",
  "MERNA14", "D3", "IL7",
  "MERNA15", "D3", "UST"
)
```

## Rename gene list

```{r, message=FALSE}
gene.list <- data.frame(entrez.id = rownames(counts))
gene.list <- gene.list %>%
  dplyr::mutate(
    SYMBOL = mapIds(org.Hs.eg.db, keys = gene.list$entrez.id, column = "SYMBOL", keytype = "ENTREZID"),
    GENETYPE = mapIds(org.Hs.eg.db, keys = gene.list$entrez.id, column = "GENETYPE", keytype = "ENTREZID"),
    ENSEMBL = mapIds(org.Hs.eg.db, keys = gene.list$entrez.id, column = "ENSEMBL", keytype = "ENTREZID")
  )

# Rename to SYMBOL
filtered.gene.list <- gene.list %>%
  dplyr::filter(
    !is.na(SYMBOL),
    !duplicated(SYMBOL))

# Rename gene IDs from ensembl to symbol
counts <- counts[rownames(counts) %in% filtered.gene.list$entrez.id, ]
rownames(counts) <- filtered.gene.list$SYMBOL[match(rownames(counts), filtered.gene.list$entrez.id)]
counts[1:4,1:4]

# Remove the ".bam" extension from colnames of counts
colnames(counts) <- sub(".bam", "", colnames(counts))

# Reorder columns of counts to match the order in coldata$SeqID
counts_run35 <- counts[, match(coldata$SeqID, colnames(counts))]

saveRDS(counts_run35, "counts_run35.rds")
```

## Generate DESeq2 object

```{r, message=FALSE, warning=FALSE}
dds <- DESeqDataSetFromMatrix(counts_run35, coldata, design = ~healthy_donor + condition)
dds <- estimateSizeFactors(dds)
idx <- rowSums(counts(dds, normalized=TRUE) >= 10 ) >= 3
dds <- dds[idx,]
dds <- DESeq(dds)
vst <- varianceStabilizingTransformation(dds)
vsn::meanSdPlot(assay(vst))
plotPCA(vst, intgroup = c("healthy_donor"))
plotPCA(vst, intgroup = c("condition"))
sizeFactors(dds)
normalized_counts <- counts(dds, normalized=TRUE)
```

### DESeq2 results - Generalised linear model (Wald test)

```{r}
res.il15 <- results(dds, contrast = c("condition","IL21","IL15"))
res.sig.il15 <- subset(res.il15, padj < 0.05)
lfc.il15 <- res.sig.il15[ order(res.sig.il15$log2FoldChange, decreasing=TRUE), ]
summary(lfc.il15)
lfc.il15
res.il2 <- results(dds,contrast = c("condition","IL21","IL2"))
res.sig.il2 <- subset(res.il2, padj < 0.05)
lfc.il2 <- res.sig.il2[ order(res.sig.il2$log2FoldChange, decreasing=TRUE), ]
summary(lfc.il2)
lfc.il2

lfc.il15.df <- lfc.il15 %>% as.data.frame() %>% rownames_to_column(.,var = 'GENE_ID')
lfc.il2.df <- lfc.il2 %>% as.data.frame() %>% rownames_to_column(.,var = 'GENE_ID')
```

## Volcano plot

```{r,fig.height=10,fig.width=15}

# Set fold change and p-value cutoffs
fold_cutoff <- 2.5
pvalue_cutoff <- 0.05

label_fold_cutoff <- 7
label_pvalue_cutoff <- 0.05

# Filter data based on cutoffs
res.il2.labels <- subset(res.il2, abs(log2FoldChange) > label_fold_cutoff & -log10(padj) > -log10(label_pvalue_cutoff))
res.il15.labels <- subset(res.il15, abs(log2FoldChange) > label_fold_cutoff & -log10(padj) > -log10(label_pvalue_cutoff))

res.il2.labels <- unique(paste0(c(rownames(res.il2.labels), "CXCL13", "IGFBP4", "TNFRSF8", "OSM", "IFNG", "DUSP6", "LTA", "SOCS2", "CSF2", "FCRL2")))
res.il15.labels <- unique(paste0(c(rownames(res.il15.labels), "CXCL13", "IGFBP4", "TNFRSF8", "OSM", "IFNG", "DUSP6", "LTA", "SOCS2", "CSF2", "FCRL2")))

filtered_data_intersect <- intersect(res.il2.labels, res.il15.labels)

# Create the first EnhancedVolcano plot for IL21 versus IL2
plot1_run35 <- EnhancedVolcano(res.il2,
    lab = rownames(res.il2),
    selectLab = filtered_data_intersect,
    drawConnectors = TRUE,
    x = 'log2FoldChange',
    y = 'padj',
    title = 'IL21 versus IL2',
    pCutoff = pvalue_cutoff,
    FCcutoff = fold_cutoff,
    pointSize = 3.0,
    boxedLabels = TRUE,
    labSize = 6.0)

# Create the second EnhancedVolcano plot for IL21 versus IL15
plot2_run35 <- EnhancedVolcano(res.il15,
    lab = rownames(res.il15),
    selectLab = filtered_data_intersect,
    drawConnectors = TRUE,
    x = 'log2FoldChange',
    y = 'padj',
    title = 'IL21 versus IL15',
    pCutoff = pvalue_cutoff,
    FCcutoff = fold_cutoff,
    pointSize = 3.0,
    boxedLabels = TRUE,
    labSize = 6.0)

# Arrange the two plots side by side using gridExtra
grid.arrange(plot1_run35, plot2_run35, ncol = 2)

```

# Run 36

```{r}
counts <- counts_run36
```

## Coldata

```{r}
coldata <- tibble::tribble(
  ~SeqID, ~healthy_donor, ~condition,
  "MERNA1", "D1", "IL2",
  "MERNA2", "D1", "IL15",
  "MERNA3", "D1", "IL21",
  "MERNA4", "D1", "IL7",
  "MERNA5", "D1", "UST",
  "MERNA6", "D2", "IL2",
  "MERNA7", "D2", "IL15",
  "MERNA8", "D2", "IL21",
  "MERNA9", "D2", "IL7",
  "MERNA10", "D2", "UST",
  "MERNA11", "D3", "IL2",
  "MERNA12", "D3", "IL15",
  "MERNA13", "D3", "IL21",
  "MERNA14", "D3", "IL7",
  "MERNA15", "D3", "UST"
)
```

## Rename gene list

```{r, message=FALSE}
gene.list <- data.frame(entrez.id = rownames(counts))
gene.list <- gene.list %>%
  dplyr::mutate(
    SYMBOL = mapIds(org.Hs.eg.db, keys = gene.list$entrez.id, column = "SYMBOL", keytype = "ENTREZID"),
    GENETYPE = mapIds(org.Hs.eg.db, keys = gene.list$entrez.id, column = "GENETYPE", keytype = "ENTREZID"),
    ENSEMBL = mapIds(org.Hs.eg.db, keys = gene.list$entrez.id, column = "ENSEMBL", keytype = "ENTREZID")
  )

# Rename to SYMBOL
filtered.gene.list <- gene.list %>%
  dplyr::filter(
    !is.na(SYMBOL),
    !duplicated(SYMBOL))

# Rename gene IDs from ensembl to symbol
counts <- counts[rownames(counts) %in% filtered.gene.list$entrez.id, ]
rownames(counts) <- filtered.gene.list$SYMBOL[match(rownames(counts), filtered.gene.list$entrez.id)]
counts[1:4,1:4]

# Remove the ".bam" extension from colnames of counts
colnames(counts) <- sub(".bam", "", colnames(counts))

# Reorder columns of counts to match the order in coldata$SeqID
counts_run36 <- counts[, match(coldata$SeqID, colnames(counts))]

saveRDS(counts_run36, "counts_run36.rds")
```

## Generate DESeq2 object

```{r, message=FALSE, warning=FALSE}
dds <- DESeqDataSetFromMatrix(counts_run36, coldata, design = ~healthy_donor + condition)
dds <- estimateSizeFactors(dds)
idx <- rowSums(counts(dds, normalized=TRUE) >= 10 ) >= 3
dds <- dds[idx,]
dds <- DESeq(dds)
vst <- varianceStabilizingTransformation(dds)
vsn::meanSdPlot(assay(vst))
plotPCA(vst, intgroup = c("healthy_donor"))
plotPCA(vst, intgroup = c("condition"))
sizeFactors(dds)
normalized_counts <- counts(dds, normalized=TRUE)
```

### DESeq2 results - Generalised linear model (Wald test)

```{r}
res.il15 <- results(dds, contrast = c("condition","IL21","IL15"))
res.sig.il15 <- subset(res.il15, padj < 0.05)
lfc.il15 <- res.sig.il15[ order(res.sig.il15$log2FoldChange, decreasing=TRUE), ]
summary(lfc.il15)
lfc.il15
res.il2 <- results(dds,contrast = c("condition","IL21","IL2"))
res.sig.il2 <- subset(res.il2, padj < 0.05)
lfc.il2 <- res.sig.il2[ order(res.sig.il2$log2FoldChange, decreasing=TRUE), ]
summary(lfc.il2)
lfc.il2

lfc.il15.df <- lfc.il15 %>% as.data.frame() %>% rownames_to_column(.,var = 'GENE_ID')
lfc.il2.df <- lfc.il2 %>% as.data.frame() %>% rownames_to_column(.,var = 'GENE_ID')

```

## Volcano plot

```{r,fig.height=10,fig.width=15}

# Set fold change and p-value cutoffs
fold_cutoff <- 2.5
pvalue_cutoff <- 0.05

label_fold_cutoff <- 7
label_pvalue_cutoff <- 0.05

# Filter data based on cutoffs
res.il2.labels <- subset(res.il2, abs(log2FoldChange) > label_fold_cutoff & -log10(padj) > -log10(label_pvalue_cutoff))
res.il15.labels <- subset(res.il15, abs(log2FoldChange) > label_fold_cutoff & -log10(padj) > -log10(label_pvalue_cutoff))

res.il2.labels <- unique(paste0(c(rownames(res.il2.labels), "CXCL13", "IGFBP4", "TNFRSF8", "OSM", "IFNG", "DUSP6", "LTA", "SOCS2", "CSF2", "FCRL2")))
res.il15.labels <- unique(paste0(c(rownames(res.il15.labels), "CXCL13", "IGFBP4", "TNFRSF8", "OSM", "IFNG", "DUSP6", "LTA", "SOCS2", "CSF2", "FCRL2")))

filtered_data_intersect <- intersect(res.il2.labels, res.il15.labels)

# Create the first EnhancedVolcano plot for IL21 versus IL2
plot1_run36 <- EnhancedVolcano(res.il2,
    lab = rownames(res.il2),
    selectLab = filtered_data_intersect,
    drawConnectors = TRUE,
    x = 'log2FoldChange',
    y = 'padj',
    title = 'IL21 versus IL2',
    pCutoff = pvalue_cutoff,
    FCcutoff = fold_cutoff,
    pointSize = 3.0,
    boxedLabels = TRUE,
    labSize = 6.0)

# Create the second EnhancedVolcano plot for IL21 versus IL15
plot2_run36 <- EnhancedVolcano(res.il15,
    lab = rownames(res.il15),
    selectLab = filtered_data_intersect,
    drawConnectors = TRUE,
    x = 'log2FoldChange',
    y = 'padj',
    title = 'IL21 versus IL15',
    pCutoff = pvalue_cutoff,
    FCcutoff = fold_cutoff,
    pointSize = 3.0,
    boxedLabels = TRUE,
    labSize = 6.0)

# Arrange the two plots side by side using gridExtra
grid.arrange(plot1_run36, plot2_run36, ncol = 2)

```

# Show all volcanoes

```{r,fig.height=20,fig.width=15}
# Create titles for the different runs
title_run35 <- textGrob("Run 35", gp = gpar(fontsize = 40, fontface = "bold"))
title_run36 <- textGrob("Run 36", gp = gpar(fontsize = 40, fontface = "bold"))

# Arrange the plots in a grid layout
grid.arrange(
    title_run35, nullGrob(), plot1_run35, plot2_run35,
    title_run36, nullGrob(), plot1_run36, plot2_run36,
    ncol = 2,
    layout_matrix = rbind(c(1, 2),
                          c(3, 4),
                          c(5, 6),
                          c(7, 8)),
    heights = c(0.1, 1, 0.1, 1)
)

```

# Sequencing saturation

The following plots show the estimated sequencing saturation for the two sequencing runs conducted for this project. The code and detailed outs are found in this directory. The code is not run in this directory because the sampling is computationally demanding; so I have ran that independently of the main markdown and just showing the final plots here.

```{r, echo=FALSE,out.width="49%",out.height="20%",fig.cap="Caption: Estimated sequencing saturation plots for run35 (left) and run36 (right)",fig.show='hold',fig.align='center', eval=TRUE}

knitr::include_graphics(c("Sequencing_saturation_outs/saturation_curve_run35.png","Sequencing_saturation_outs/saturation_curve_run36.png"))
``` 

# Conclusion

The sequencing runs are called 35 and 36 because this is the numbers given by the sequencing hub, and have just used this code as shorthand for describing the separate sequencing runs. Go to the HPC analysis script for the long sequencing name.

The plots for run 35 and run 36 are very similar. They both reach random sampling sequencing saturation. The technical batch effect will be greater than any extra information by combining the sequencing runs. Therefore, the final analysis will just contain sequencing run 35.

The final analysis for the 2024 paper is the same as is shown in the 'run 35' chunks - with IL-7 removed as this was not necessary for the paper.
