---
title: "Untitled"
author: "D Griffiths"
date: "`r Sys.Date()`"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Process raw sequencing files

## Get data

Data is stored on the RDS under 'raw data'

Transfer to the HPC for processing

Inflate the compressed files (and can run the md5 checksum to ensure that the transfer was successful)

Then move on to processing using nextflow. A tutorial can be found here:

<https://github.com/dgrfs/NFCore_RNASeq_Cardiff_University_Hawk>

``` sh
# lang: sh

tar -xvf 161006_Run35_fastq.tar.gz
```

Add sequencing run ID to the files

-   for run35 - AH2572BGXY

-   for run36 - AH33YJBGXY

``` sh
# lang: sh

for file in *.fastq.gz; do
    mv "$file" "${file%.fastq.gz}_AH2572BGXY.fastq.gz"
done

for file in *.fastq.gz; do
    mv "$file" "${file%.fastq.gz}_AH33YJBGXY.fastq.gz"
done
```

``` r
# lang: r

library(tidyverse)
files <- list.files(pattern = ".fastq.gz")

df <- tibble(filename = files)
df$file_to_split <- df$filename
df <- df %>% separate(file_to_split, c("sample", "SeqID", "Read", "lane", "seq.run", "Extension"), extra = "merge")

# Pivot the data
df_wide <- df %>%
  pivot_wider(names_from = Read, values_from = filename, names_prefix = "fastq_") %>%
  rename(fastq_1 = fastq_R1, fastq_2 = fastq_R2)
df_wide$strandedness <- rep("auto", length(df_wide$sample))
# select names
df_final <- df_wide %>%
  select(c(sample, fastq_1, fastq_2, strandedness, seq.run, lane))
  
# Write the output data frame to a CSV file
write.csv(df_final, "samplesheet.csv", row.names = FALSE)
```

``` sh
# lang: sh

mkdir ref
cd ref

cat << 'EOF' > get-refs.sh
 #!/bin/bash
 VERSION=108
 wget -L ftp://ftp.ensembl.org/pub/release-$VERSION/fasta/homo_sapiens/dna/Homo_sapiens.GRCh38.dna_sm.primary_assembly.fa.gz
 wget -L ftp://ftp.ensembl.org/pub/release-$VERSION/gtf/homo_sapiens/Homo_sapiens.GRCh38.$VERSION.gtf.gz
EOF

bash get-refs.sh

# back to working directory
cd ..

# get config
wget https://raw.githubusercontent.com/dgrfs/NFCore_RNASeq_Cardiff_University_Hawk/main/adapted.scw.config

# get params
wget https://raw.githubusercontent.com/dgrfs/NFCore_RNASeq_Cardiff_University_Hawk/main/params.yaml

# set up screen/tmux to run in background
module load tmux
tmux

# load dependencies
module load singularity-ce/3.11.4
module load nextflow/22.10.6

nextflow run nf-core/rnaseq \
    -r 3.12.0 \
    -params-file params.yaml \
    -profile singularity \
    -c adapted.scw.config \
    --fasta $PWD/ref/Homo_sapiens.GRCh38.dna_sm.primary_assembly.fa.gz \
    --gtf $PWD/ref/Homo_sapiens.GRCh38.108.gtf.gz 
```
